internal/dataplane/kong_client.go

```diff
@@ -451,14 +466,31 @@ func (c *KongClient) triggerKubernetesObjectReport(objs ...client.Object) {
 	// control-plane can update the Kubernetes object statuses for affected objs.
 	// this has to be done in a separate loop so that the filter is in place
 	// before the objects are enqueued, as the filter is used by the control-plane
-	for _, obj := range objs {
-		c.kubernetesObjectStatusQueue.Publish(obj)
+
+	// create a new set to de-duplicate the objects to avoid duplicate events sent to controllers.
+	set = k8sobj.ConfigurationStatusSet{}
+	for _, obj := range reportedObjects {
+		if set.Get(obj) == k8sobj.ConfigurationStatusUnknown {
+			set.Insert(obj, true)
+			c.kubernetesObjectStatusQueue.Publish(obj)
+		}
+	}
+
+	for _, translationFailure := range translationFailures {
+		for _, obj := range translationFailure.CausingObjects() {
+			// We only check if the object has been added to the set to avoid multiple events on the same object.
+			// Controllers will get configuration status from c.kubernetesObjectReportsFilter.
+			if set.Get(obj) == k8sobj.ConfigurationStatusUnknown {
+				set.Insert(obj, false)
+				c.kubernetesObjectStatusQueue.Publish(obj)
+			}
+		}
 	}
```

Comment:
Although it undoubtedly works, I must admit found it hard to understand at first glance. Do you think we could avoid explaining the trickery in comments and instead have that extracted to a separate function? I would also say that we might not want to reuse `ConfigurationStatusSet` here as it's a little bit different purpose than the one it was built for, which might be misleading to the reader. Maybe something like following? 

```suggestion
	for _, obj := range uniqueObjects(reportedObjects, translationFailures) {
		c.kubernetesObjectStatusQueue.Publish(obj)
	}
}

func uniqueObjects(reportedObjects []client.Object, resourceFailures []failures.ResourceFailure) []client.Object {
	allCausingObjects := lo.FlatMap(resourceFailures, func(f failures.ResourceFailure, _ int) []client.Object {
		return f.CausingObjects()
	})

	allObjects := append(reportedObjects, allCausingObjects...)
	return lo.UniqBy(allObjects, func(obj client.Object) string {
		return obj.GetObjectKind().GroupVersionKind().String() + "/" +
			obj.GetNamespace() + "/" + obj.GetName()
	})
}
```

`uniqueObjects` additionally could be even unit tested, WDYT?

Reply:
sounds good, will do.

Reply:
done